### PR TITLE
hw-registration-validation implementation by fetis_off

### PR DESCRIPTION
**Pull request by fetis-off**

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -2,13 +2,31 @@
...
+        }
+        if (user.getPassword() == null
+                || user.getPassword().length() < MIN_LOG_PASSWORD_LENGTH) {
+            throw new InvalidDataException("Password must be at least 6 characters");
+        }
+        if (user.getAge() < MIN_AGE) {
+            throw new InvalidDataException("Age must be at least 18 years old");
+        }
+        storageDao.add(user);
+        return user;
```
Comment (MentorMateAcademy) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:30]: divide for private methods that check login, age, etc

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    @BeforeEach
+    void setUp() {
+        storage = new StorageDaoImpl();
+        service = new RegistrationServiceImpl();
+    }
+
+    @Test
+    void registerExistingLogin_notOk() {
+        User existingUser = new User();
+        existingUser.setLogin("TESTLOGIN");
```
Comment (MentorMateAcademy) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:26]: save values to constant and reuse it

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/model/InvalidDataException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.model;
+
+public class InvalidDataException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/model/InvalidDataException.java:3]:
```suggestion
public class RegistrationException extends RuntimeException {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        assertEquals("Password must be at least 6 characters", exception.getMessage());
+    }
+
+    @Test
+    void registerValidUser_Ok() {
+        User testUser = new User();
+        testUser.setLogin(TEST_LOGIN_ONE);
+        testUser.setPassword(TEST_PASSWORD);
+        testUser.setAge(25);
+        assertDoesNotThrow(() -> service.register(testUser));
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:66]:
```suggestion
        User actual = service.register(testUser));
        assertEquals(testUser, actual);
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        testUser.setPassword(null);
+        testUser.setAge(25);
+        InvalidDataException exception = assertThrows(InvalidDataException.class, () -> {
+            service.register(testUser);
+        });
+        assertEquals("Password must be at least 6 characters", exception.getMessage());
+    }
+
+    @Test
+    void registerWithNumbersPasswordAndLogin_Ok() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:105]: Actually, you don't have any logic about numbers or symbols in login or password, so this test case os redundant. 
It's better to add test case for negative age

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 4 above)
Comment (Elena-Bruyako) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:66]: In good case, we should check assertEquals or assertTrue (positive case), not assertDoesNotThrow

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 6]
```diff
@@ -2,13 +2,47 @@
...
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final byte MIN_LOG_PASSWORD_LENGTH = 6;
+    private static final byte MIN_AGE = 18;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        checkSameLogin(user);
+        checkLogin(user);
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:16]: It's better to combine into one method checkLogin
Also, it's better to check if user == null first

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by Elena-Bruyako — APPROVED** [review]
